grobid-core/src/main/java/org/grobid/core/engines/PdfStructureParser.java

```diff
@@ -0,0 +1,846 @@
+package org.grobid.core.engines;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.tuple.Pair;
+import org.grobid.core.GrobidModels;
+import org.grobid.core.data.*;
+import org.grobid.core.document.*;
+import org.grobid.core.engines.citations.LabeledReferenceResult;
+import org.grobid.core.engines.label.SegmentationLabels;
+import org.grobid.core.engines.label.TaggingLabel;
+import org.grobid.core.engines.label.TaggingLabels;
+import org.grobid.core.layout.LayoutToken;
+import org.grobid.core.layout.LayoutTokenization;
+import org.grobid.core.layout.Page;
+import org.grobid.core.tokenization.TaggingTokenCluster;
+import org.grobid.core.tokenization.TaggingTokenClusteror;
+import org.grobid.core.utilities.LayoutTokensUtil;
+import org.grobid.core.utilities.TextUtilities;
+import org.grobid.core.utilities.matching.EntityMatcherException;
+import org.grobid.core.utilities.matching.ReferenceMarkerMatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.regex.Pattern;
+
+
+/**
+ * Uses Grobid's mark-up to produce detailed token-level mark-up of the PDF text
+ * in a form appropriate for consumption by the PDF Structure Service
+ * https://github.com/allenai/s2-pdf-structure-service
+ */
+public class PdfStructureParser {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PdfStructureParser.class);
+
+    public PdfStructure extractStructure(Engine engine, Document doc) {
```

Comment:
This is a non-intrusive examination of the raw Grobid annotations. It works very similarly to the `TEIFormatter` but builds a `PdfStructure` class instead of an XML document.